# Finding and cleaning Data

In [2]:
#import libraries
import pandas as pd
import numpy as np
import os

In [12]:
# Creating a folder path
path = r'C:\Users\Lenovo\Desktop\Career Foundry\Achievement 6'

In [14]:
#Using the os.path.function
df = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'Covid.csv'), index_col = False)


In [16]:
#Checking the data:
df.head(10)

Data As Of  Start Date    End Date     Group  Year  Month          State  \
0  09/24/2023  01/01/2020  09/23/2023  By Total   NaN    NaN  United States   
1  09/24/2023  01/01/2020  09/23/2023  By Total   NaN    NaN  United States   
2  09/24/2023  01/01/2020  09/23/2023  By Total   NaN    NaN  United States   
3  09/24/2023  01/01/2020  09/23/2023  By Total   NaN    NaN  United States   
4  09/24/2023  01/01/2020  09/23/2023  By Total   NaN    NaN  United States   
5  09/24/2023  01/01/2020  09/23/2023  By Total   NaN    NaN  United States   
6  09/24/2023  01/01/2020  09/23/2023  By Total   NaN    NaN  United States   
7  09/24/2023  01/01/2020  09/23/2023  By Total   NaN    NaN  United States   
8  09/24/2023  01/01/2020  09/23/2023  By Total   NaN    NaN  United States   
9  09/24/2023  01/01/2020  09/23/2023  By Total   NaN    NaN  United States   

        Condition Group                Condition ICD10_codes   Age Group  \
0  Respiratory diseases  Influenza and pneumonia     J09-J18        0-24   
1  Respiratory diseases  Influenza and pneumonia     J09-J18       25-34   
2  Respiratory diseases  Influenza and pneumonia     J09-J18       35-44   
3  Respiratory diseases  Influenza and pneumonia     J09-J18       45-54   
4  Respiratory diseases  Influenza and pneumonia     J09-J18       55-64   
5  Respiratory diseases  Influenza and pneumonia     J09-J18       65-74   
6  Respiratory diseases  Influenza and pneumonia     J09-J18       75-84   
7  Respiratory diseases  Influenza and pneumonia     J09-J18         85+   
8  Respiratory diseases  Influenza and pneumonia     J09-J18  Not stated   
9  Respiratory diseases  Influenza and pneumonia     J09-J18    All Ages   

   COVID-19 Deaths  Number of Mentions Flag  
0           1569.0              1647.0  NaN  
1           5804.0              6029.0  NaN  
2          15080.0             15699.0  NaN  
3          37414.0             38878.0  NaN  
4          82668.0             85708.0  NaN  
5         129005.0            133088.0  NaN  
6         138503.0            141868.0  NaN  
7         121119.0            123018.0  NaN  
8             12.0                12.0  NaN  
9         531174.0            545947.0  NaN

In [18]:
# Checking for total rows and columns
df.shape

(621000, 14)

In [20]:
# No one will understand scientific numbers. Changing the display setting:
pd.set_option('display.float_format', lambda x: '%.2f' % x)

In [22]:
# Check the basic statistics of the dataset
df.describe()

Year     Month  COVID-19 Deaths  Number of Mentions
count 608580.00 558900.00        437551.00           443423.00
mean    2021.41      6.20           120.12              129.33
std        1.09      3.35          2980.20             3203.94
min     2020.00      1.00             0.00                0.00
25%     2020.00      3.00             0.00                0.00
50%     2021.00      6.00             0.00                0.00
75%     2022.00      9.00            18.00               19.00
max     2023.00     12.00       1146242.00          1146242.00

## Cleaning the data

In [25]:
#Checking for missing values:
df.isnull().sum()

Data As Of                 0
Start Date                 0
End Date                   0
Group                      0
Year                   12420
Month                  62100
State                      0
Condition Group            0
Condition                  0
ICD10_codes                0
Age Group                  0
COVID-19 Deaths       183449
Number of Mentions    177577
Flag                  437551
dtype: int64

In [27]:
# Check the data-types of the dataframe
df.dtypes

Data As Of             object
Start Date             object
End Date               object
Group                  object
Year                  float64
Month                 float64
State                  object
Condition Group        object
Condition              object
ICD10_codes            object
Age Group              object
COVID-19 Deaths       float64
Number of Mentions    float64
Flag                   object
dtype: object

### Checking Each Column Separately to understand the reason for missing values

In [30]:
#YEAR Frequency:
df['Year'].value_counts(dropna = False)

Year
2020.00    161460
2021.00    161460
2022.00    161460
2023.00    124200
NaN         12420
Name: count, dtype: int64

In [32]:
#Month Frequency:
df['Month'].value_counts(dropna = False)

Month
NaN      62100
1.00     49680
2.00     49680
3.00     49680
4.00     49680
5.00     49680
6.00     49680
7.00     49680
8.00     49680
9.00     49680
10.00    37260
11.00    37260
12.00    37260
Name: count, dtype: int64

In [34]:
#COVID-19 Deaths Frequency
df['COVID-19 Deaths'].value_counts(dropna = False)

COVID-19 Deaths
0.00        288824
NaN         183449
10.00         6438
11.00         5731
12.00         5112
             ...  
9837.00          1
15731.00         1
12011.00         1
11086.00         1
6485.00          1
Name: count, Length: 4668, dtype: int64

In [36]:
# Number of mentions frequency:
df['Number of Mentions'].value_counts(dropna = False)

Number of Mentions
0.00        288824
NaN         177577
10.00         6517
11.00         5904
12.00         5198
             ...  
25785.00         1
12047.00         1
3652.00          1
42749.00         1
8794.00          1
Name: count, Length: 4860, dtype: int64

In [38]:
# Flag frequency:
df['Flag'].value_counts(dropna = False)

Flag
NaN                                                                                                                           437551
One or more data cells have counts between 1-9 and have been suppressed in accordance with NCHS confidentiality standards.    183449
Name: count, dtype: int64

### Interesting insights:
1.	The 0 in Number of mentions and in Number of deaths is the same: 288824
2.	The missing Values in Covid – 19 deaths is same as Suppressed data: 183449


### Checking some rows to understand better:

In [48]:
# Find rows where "Covid-19 Deaths" column has missing values
missing_values_df = df[df['COVID-19 Deaths'].isna()]

# Display the first 5 rows with missing values in "Covid-19 Deaths"
print(missing_values_df.head(5))

     Data As Of  Start Date    End Date     Group  Year  Month    State  \
241  09/24/2023  01/01/2020  09/23/2023  By Total   NaN    NaN  Alabama   
250  09/24/2023  01/01/2020  09/23/2023  By Total   NaN    NaN  Alabama   
270  09/24/2023  01/01/2020  09/23/2023  By Total   NaN    NaN  Alabama   
280  09/24/2023  01/01/2020  09/23/2023  By Total   NaN    NaN  Alabama   
301  09/24/2023  01/01/2020  09/23/2023  By Total   NaN    NaN  Alabama   

          Condition Group                                 Condition  \
241  Respiratory diseases        Chronic lower respiratory diseases   
250  Respiratory diseases       Adult respiratory distress syndrome   
270  Respiratory diseases                        Respiratory arrest   
280  Respiratory diseases  Other diseases of the respiratory system   
301  Circulatory diseases                    Ischemic heart disease   

                                           ICD10_codes Age Group  \
241                                            J40-J47

### Checking other columns for more insight

In [51]:
# State frequency:
df['State'].value_counts(dropna = False)

State
United States           11500
Pennsylvania            11500
Nevada                  11500
New Hampshire           11500
New Jersey              11500
New Mexico              11500
New York                11500
New York City           11500
North Carolina          11500
North Dakota            11500
Ohio                    11500
Oklahoma                11500
Oregon                  11500
Rhode Island            11500
Alabama                 11500
South Carolina          11500
South Dakota            11500
Tennessee               11500
Texas                   11500
Utah                    11500
Vermont                 11500
Virginia                11500
Washington              11500
West Virginia           11500
Wisconsin               11500
Wyoming                 11500
Nebraska                11500
Montana                 11500
Missouri                11500
Mississippi             11500
Alaska                  11500
Arizona                 11500
Arkansas                11500
Cali

In [53]:
# Condition Group frequency:
df['Condition Group'].value_counts(dropna = False)

Condition Group
Circulatory diseases                                                         189000
Respiratory diseases                                                         162000
Sepsis                                                                        27000
Malignant neoplasms                                                           27000
Diabetes                                                                      27000
Obesity                                                                       27000
Alzheimer disease                                                             27000
Vascular and unspecified dementia                                             27000
Renal failure                                                                 27000
Intentional and unintentional injury, poisoning, and other adverse events     27000
All other conditions and causes (residual)                                    27000
COVID-19                                                    

In [55]:
#Group Frequency"
df['Group'].value_counts(dropna = False)

Group
By Month    558900
By Year      49680
By Total     12420
Name: count, dtype: int64

In [61]:
#Dropping Group Column as it does not provide any valuable input:
df_dropped = df.drop('Group', axis=1)  # axis=1 indicates we are dropping a column

In [63]:
#Check for New DF:
df_dropped.dtypes

Data As Of             object
Start Date             object
End Date               object
Year                  float64
Month                 float64
State                  object
Condition Group        object
Condition              object
ICD10_codes            object
Age Group              object
COVID-19 Deaths       float64
Number of Mentions    float64
Flag                   object
dtype: object

In [65]:
# Drop rows where the "State" column is "United States" or "Puerto Rico"
df_filtered = df[~df['State'].isin(['United States', 'Puerto Rico'])]

### Imputing missing values in Death and No. of Mentions Column between 0 and 9

In [68]:
# Generate random integers between 0 and 9 to fill missing values
random_numbers = np.random.randint(0, 10, df_filtered['COVID-19 Deaths'].isna().sum())

In [70]:
# Impute missing values in "COVID-19 Deaths" column with random numbers between 0 and 9
df_filtered.loc[df_filtered['COVID-19 Deaths'].isna(), 'COVID-19 Deaths'] = random_numbers

In [76]:
# Generate random integers between 0 and 9 to fill missing values
random_numbers2 = np.random.randint(0, 10, df_filtered['Number of Mentions'].isna().sum())

In [78]:
# Impute missing values in "Number of mentions" column with random numbers between 0 and 9
df_filtered.loc[df_filtered['Number of Mentions'].isna(), 'Number of Mentions'] = random_numbers2

In [82]:
#Age group
df_filtered['Age Group'].value_counts(dropna = False)

Age Group
0-24          59800
25-34         59800
35-44         59800
45-54         59800
55-64         59800
65-74         59800
75-84         59800
85+           59800
Not stated    59800
All Ages      59800
Name: count, dtype: int64

In [86]:
#Removing Rows in "Age Group" Column Not not stated and All Ages as they will not be valuable
df_cleaned = df_filtered[df_filtered['Age Group'] != 'All Ages']


In [88]:
df_cleaned_covid = df_cleaned[df_cleaned['Age Group'] != 'Not Stated']

In [90]:
#Checking for missing values:
df_cleaned_covid.isnull().sum()

Data As Of                 0
Start Date                 0
End Date                   0
Group                      0
Year                   10764
Month                  53820
State                      0
Condition Group            0
Condition                  0
ICD10_codes                0
Age Group                  0
COVID-19 Deaths            0
Number of Mentions         0
Flag                  374669
dtype: int64

In [92]:
#Removing rows with missing values in Year and month Column:
df_cleaned = df_cleaned_covid.dropna(subset=['Year', 'Month'])

In [94]:
df_cleaned.dtypes

Data As Of             object
Start Date             object
End Date               object
Group                  object
Year                  float64
Month                 float64
State                  object
Condition Group        object
Condition              object
ICD10_codes            object
Age Group              object
COVID-19 Deaths       float64
Number of Mentions    float64
Flag                   object
dtype: object

### Checking for mixed data types:

In [97]:
for col in df_cleaned.columns.tolist():
    weird = (df_cleaned[[col]].map(type) != df_cleaned[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_cleaned[weird]) > 0:
        print ("Mixed-type data found in column: {col}")
    else: 
        print("No mixed-type data in column: {col}")

No mixed-type data in column: {col}
No mixed-type data in column: {col}
No mixed-type data in column: {col}
No mixed-type data in column: {col}
No mixed-type data in column: {col}
No mixed-type data in column: {col}
No mixed-type data in column: {col}
No mixed-type data in column: {col}
No mixed-type data in column: {col}
No mixed-type data in column: {col}
No mixed-type data in column: {col}
No mixed-type data in column: {col}
No mixed-type data in column: {col}
Mixed-type data found in column: {col}


In [101]:
#Changing the data types 
df_cleaned.loc[:, 'Data As Of'] = pd.to_datetime(df_cleaned['Data As Of'])
df_cleaned.loc[:, 'Start Date'] = pd.to_datetime(df_cleaned['Start Date'])
df_cleaned.loc[:, 'End Date'] = pd.to_datetime(df_cleaned['End Date'])

In [109]:
## Convert categorical columns to category type
categorical_columns = ['State', 'Condition Group', 'Condition', 'ICD10_codes', 'Age Group', 'Flag']
for column in categorical_columns:
    df_cleaned.loc[:, column] = df_cleaned[column].astype('category')

### Used the loc function to avoid copyright warning

In [112]:
df_cleaned.dtypes

Data As Of            datetime64[ns]
Start Date            datetime64[ns]
End Date              datetime64[ns]
Group                         object
Year                         float64
Month                        float64
State                       category
Condition Group             category
Condition                   category
ICD10_codes                 category
Age Group                   category
COVID-19 Deaths              float64
Number of Mentions           float64
Flag                        category
dtype: object

In [115]:
#Dropping Group Column as it does not provide any valuable input:
df_cleaned = df_cleaned.drop('Group', axis=1)  # axis=1 indicates we are dropping a column

In [117]:
#Check again for missing values:
df_cleaned.isnull().sum()

Data As Of                 6
Start Date                 6
End Date                   6
Year                       6
Month                      6
State                      6
Condition Group            6
Condition                  6
ICD10_codes                6
Age Group                  6
COVID-19 Deaths            6
Number of Mentions         6
Flag                  333472
dtype: int64

In [119]:
df_cleaned = df_cleaned.dropna(subset=['Data As Of', 'Start Date'])

In [121]:
#Check again for missing values:
df_cleaned.isnull().sum()

Data As Of                 0
Start Date                 0
End Date                   0
Year                       0
Month                      0
State                      0
Condition Group            0
Condition                  0
ICD10_codes                0
Age Group                  0
COVID-19 Deaths            0
Number of Mentions         0
Flag                  333466
dtype: int64

In [123]:
df_cleaned.describe()

Data As Of                     Start Date  \
count               484380                         484380   
mean   2023-09-24 00:00:00  2021-10-31 07:28:00.000000768   
min    2023-09-24 00:00:00            2020-01-01 00:00:00   
25%    2023-09-24 00:00:00            2020-12-01 00:00:00   
50%    2023-09-24 00:00:00            2021-11-01 00:00:00   
75%    2023-09-24 00:00:00            2022-10-01 00:00:00   
max    2023-09-24 00:00:00            2023-09-01 00:00:00   
std                    NaN                            NaN   

                            End Date      Year     Month  COVID-19 Deaths  \
count                         484380 484380.00 484380.00        484380.00   
mean   2021-11-29 13:51:59.999999488   2021.40      6.20             9.55   
min              2020-01-31 00:00:00   2020.00      1.00             0.00   
25%              2020-12-31 00:00:00   2020.00      3.00             0.00   
50%              2021-11-30 00:00:00   2021.00      6.00             0.00   
75%              2022-10-31 00:00:00   2022.00      9.00             6.00   
max              2023-09-23 00:00:00   2023.00     12.00          5094.00   
std                              NaN      1.08      3.35            51.01   

       Number of Mentions  
count           484380.00  
mean                10.32  
min                  0.00  
25%                  0.00  
50%                  0.00  
75%                  6.00  
max               5094.00  
std                 53.97

In [125]:
# State frequency:
df_cleaned['State'].value_counts(dropna = False)

State
Alabama                 9315
Alaska                  9315
Nevada                  9315
New Hampshire           9315
New Jersey              9315
New Mexico              9315
New York                9315
New York City           9315
North Carolina          9315
North Dakota            9315
Ohio                    9315
Oklahoma                9315
Oregon                  9315
Pennsylvania            9315
Rhode Island            9315
South Carolina          9315
South Dakota            9315
Tennessee               9315
Texas                   9315
Utah                    9315
Vermont                 9315
Virginia                9315
Washington              9315
West Virginia           9315
Wisconsin               9315
Nebraska                9315
Montana                 9315
Missouri                9315
Idaho                   9315
Arizona                 9315
Arkansas                9315
California              9315
Colorado                9315
Connecticut             9315
Delaware

### Check for duplicates

In [128]:
# Check for duplicate rows in the dataset
df_cleaned.duplicated().sum()

0

In [130]:
df_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
Index: 484380 entries, 72450 to 610604
Data columns (total 13 columns):
 #   Column              Non-Null Count   Dtype         
---  ------              --------------   -----         
 0   Data As Of          484380 non-null  datetime64[ns]
 1   Start Date          484380 non-null  datetime64[ns]
 2   End Date            484380 non-null  datetime64[ns]
 3   Year                484380 non-null  float64       
 4   Month               484380 non-null  float64       
 5   State               484380 non-null  category      
 6   Condition Group     484380 non-null  category      
 7   Condition           484380 non-null  category      
 8   ICD10_codes         484380 non-null  category      
 9   Age Group           484380 non-null  category      
 10  COVID-19 Deaths     484380 non-null  float64       
 11  Number of Mentions  484380 non-null  float64       
 12  Flag                150914 non-null  category      
dtypes: category(6), datetime64[ns]

In [132]:
# Compressing datatypes for saving space:

df_cleaned['Year'] = df_cleaned['Year'].astype('float32')
df_cleaned['Month'] = df_cleaned['Month'].astype('float32')
df_cleaned['COVID-19 Deaths'] = df_cleaned['COVID-19 Deaths'].astype('float32')
df_cleaned['Number of Mentions'] = df_cleaned['Number of Mentions'].astype('float32')

In [134]:
df_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
Index: 484380 entries, 72450 to 610604
Data columns (total 13 columns):
 #   Column              Non-Null Count   Dtype         
---  ------              --------------   -----         
 0   Data As Of          484380 non-null  datetime64[ns]
 1   Start Date          484380 non-null  datetime64[ns]
 2   End Date            484380 non-null  datetime64[ns]
 3   Year                484380 non-null  float32       
 4   Month               484380 non-null  float32       
 5   State               484380 non-null  category      
 6   Condition Group     484380 non-null  category      
 7   Condition           484380 non-null  category      
 8   ICD10_codes         484380 non-null  category      
 9   Age Group           484380 non-null  category      
 10  COVID-19 Deaths     484380 non-null  float32       
 11  Number of Mentions  484380 non-null  float32       
 12  Flag                150914 non-null  category      
dtypes: category(6), datetime64[ns]

In [136]:
#Saving cleaned data for further analysis:
# Exporting "df_covid_final_cleaned" dataframe to CSV
df_cleaned.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'Covid_clean.csv'), index=False)